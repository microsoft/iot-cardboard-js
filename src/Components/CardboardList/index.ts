export * from './CardboardList';
export * from './CardboardListItem';
