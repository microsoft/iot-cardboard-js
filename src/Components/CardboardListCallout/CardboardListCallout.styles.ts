import {
    IButtonStyles,
    ICalloutContentStyles,
    memoizeFunction,
    mergeStyleSets,
    Theme
} from '@fluentui/react';
import { CardboardClassNamePrefix } from '../../Models/Constants';

const classPrefix = `${CardboardClassNamePrefix}-list-callout`;
const classNames = {
    title: `${classPrefix}-title`,
    list: `${classPrefix}-list`,
    resultText: `${classPrefix}-resultText`
};
export const getCardboardListCalloutComponentStyles = memoizeFunction(() => {
    const ellipseStyles = {
        overflow: 'hidden',
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis'
    };
    return mergeStyleSets({
        title: [
            classNames.title,
            {
                marginTop: 0,
                fontWeight: 600,
                ...ellipseStyles
            }
        ],
        list: [
            classNames.list,
            {
                paddingTop: 12,
                '.ms-List-surface': {
                    maxHeight: 200,
                    overflowX: 'hidden',
                    overflowY: 'auto'
                },
                '.cb-basic-list-item-root': {
                    padding: '8px 12px 8px 0'
                },
                '.ms-List-cell:last-of-type': {
                    '.cb-basic-list-item-root': {
                        paddingBottom: 0
                    }
                }
            }
        ],
        resultText: [
            classNames.resultText,
            {
                fontSize: 12,
                marginTop: 8,
                opacity: 0.6
            }
        ]
    });
});

export const getCardboardListCalloutStyles = memoizeFunction(
    (theme: Theme) =>
        ({
            root: {
                padding: '16px 20px 20px',
                width: 300,
                backgroundColor: theme.semanticColors.bodyBackground
            },
            calloutMain: {
                backgroundColor: 'unset'
            }
        } as ICalloutContentStyles)
);

export const cardboardListCalloutPrimaryButtonStyles = {
    root: {
        marginTop: 16
    }
} as IButtonStyles;
