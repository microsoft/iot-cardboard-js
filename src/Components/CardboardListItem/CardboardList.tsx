import {
    FocusZone,
    FocusZoneDirection,
    IFocusZoneProps,
    IListProps,
    List
} from '@fluentui/react';
import React, { ReactNode } from 'react';
import { CardboardListItem, CardboardListItemProps } from './CardboardListItem';

export interface ICardboardListProps<T> {
    /** unique identifier for this list of items. Will be joined with index */
    listKey: string;
    /** optional prop to set any specific focus zone props needed for special cases */
    focusZoneProps?: IFocusZoneProps;
    /** callback fired for each item in the list to generate the internal properties for the list items */
    getListItemProps: (item: T, index: number) => CardboardListItemProps<T>;
    /** Collection of items to include in the list */
    items: T[];
    /** optional prop to set any specific list props needed for special cases */
    listProps?: IListProps;
    /** text to highlight on the primary text. mainly used for indicating search matches */
    textToHighlight?: string;
}

export const CardboardList = <T extends unknown>({
    focusZoneProps,
    getListItemProps,
    items,
    listKey,
    listProps,
    textToHighlight
}: ICardboardListProps<T> & { children?: ReactNode }) => {
    return (
        <>
            <FocusZone
                {...focusZoneProps}
                direction={FocusZoneDirection.vertical}
            >
                <List
                    {...listProps}
                    items={items}
                    data-testid={`cardboard-list-${listKey}`}
                    onRenderCell={(item, index) => (
                        <CardboardListItem
                            {...getListItemProps(item, index)}
                            listKey={listKey}
                            index={index}
                            item={item}
                            textToHighlight={textToHighlight}
                        />
                    )}
                />
            </FocusZone>
        </>
    );
};
