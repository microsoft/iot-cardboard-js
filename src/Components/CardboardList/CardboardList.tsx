import { FocusZone, FocusZoneDirection, List } from '@fluentui/react';
import React, { ReactNode } from 'react';
import {
    ICardboardListItemPropsInternal,
    ICardboardListProps
} from './CardboardList.types';
import { CardboardListItem } from './CardboardListItem';

export const CardboardList = <T extends unknown>({
    focusZoneProps,
    items,
    listKey,
    listProps,
    textToHighlight
}: ICardboardListProps<T> & { children?: ReactNode }) => {
    const onRenderListItem = (
        item: ICardboardListItemPropsInternal<T>,
        index: number
    ) => (
        <CardboardListItem<T>
            {...item}
            listKey={listKey}
            index={index}
            textToHighlight={textToHighlight}
        />
    );

    return (
        <>
            <FocusZone
                {...focusZoneProps}
                direction={FocusZoneDirection.vertical}
            >
                <List
                    {...listProps}
                    data-testid={`cardboard-list-${listKey}`}
                    items={items}
                    onRenderCell={onRenderListItem}
                />
            </FocusZone>
        </>
    );
};
