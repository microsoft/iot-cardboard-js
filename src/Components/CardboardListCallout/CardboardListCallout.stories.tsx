import React from 'react';
import { ComponentStory } from '@storybook/react';
import { userEvent, within } from '@storybook/testing-library';
import {
    getDefaultStoryDecorator,
    sleep
} from '../../Models/Services/StoryUtilities';
import { ICardboardListCalloutProps } from './CardboardListCallout.types';
import CardboardListCallout from './CardboardListCallout';
import { ICardboardListItem } from '../CardboardList/CardboardList.types';
import { ActionButton, DirectionalHint } from '@fluentui/react';
import { useBoolean, useId } from '@fluentui/react-hooks';

const componentStyle = {
    height: '600px',
    width: '300px'
};

export default {
    title: 'Components/CardboardListCallout',
    component: CardboardListCallout,
    decorators: [
        getDefaultStoryDecorator<ICardboardListCalloutProps<string>>(
            componentStyle
        )
    ]
};

// fake data to be used for either CardboardList or CardboardBasicList
const defaultCardboardBasicListItems = [
    'Item 1',
    'Item 2',
    'Item 3',
    'Item 4',
    'Item 5'
];
interface IFakeCardboardListItem {
    itemId: string;
    itemDescription: string | undefined;
}
const defaultCardboardListItems: IFakeCardboardListItem[] = [
    {
        itemId: 'Item 1',
        itemDescription: 'description for item 1'
    },
    {
        itemId: 'Item 2',
        itemDescription: 'description for item 1'
    },
    {
        itemId: 'Item 3',
        itemDescription: 'description for item 3'
    },
    {
        itemId: 'Item 4',
        itemDescription: 'description for item 4'
    },
    {
        itemId: 'Item 5',
        itemDescription: 'description for item 5'
    }
];

const defaultOnClickHandler = (item) => {
    alert(`clicked item ${item.itemId}`);
};

const getDefaultCardboardListItems = (): ICardboardListItem<IFakeCardboardListItem>[] =>
    defaultCardboardListItems.map((item) => ({
        ariaLabel: '',
        iconStart: { name: 'Shapes' },
        iconEnd: { name: 'Add' },
        item: item,
        onClick: defaultOnClickHandler,
        textPrimary: item.itemId,
        textSecondary: item.itemDescription
    }));

const getCommonDefaultProps = (): Partial<ICardboardListCalloutProps> => ({
    title: 'Example title',
    listKey: 'testList',
    directionalHint: DirectionalHint.bottomLeftEdge,
    filterPlaceholder: 'Search',
    noResultText: 'No result'
});

type TemplateStory = ComponentStory<typeof CardboardListCallout>;
const Template = (args) => {
    const testCalloutTargetId = useId('test-cardboard-list-callout-target');
    const [isCalloutVisible, { toggle: toggleIsCalloutVisible }] = useBoolean(
        false
    );

    return (
        <>
            <ActionButton
                data-testid="showCalloutButton"
                id={testCalloutTargetId}
                text="Show list callout"
                onClick={toggleIsCalloutVisible}
            />
            {isCalloutVisible && (
                <CardboardListCallout
                    {...args}
                    calloutTarget={testCalloutTargetId}
                    onDismiss={toggleIsCalloutVisible}
                />
            )}
        </>
    );
};

export const ExampleCardboardListCallout = Template.bind({}) as TemplateStory;
ExampleCardboardListCallout.args = {
    ...getCommonDefaultProps(),
    listType: 'Complex',
    listItems: getDefaultCardboardListItems(),
    filterPredicate: (fakeItem: IFakeCardboardListItem, searchTerm) =>
        fakeItem.itemId.toLowerCase().includes(searchTerm.toLowerCase())
};
ExampleCardboardListCallout.play = async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    // click on target to show callout
    const showCalloutButton = await canvas.findByTestId('showCalloutButton');
    await userEvent.click(showCalloutButton);

    // wait until callout show animation to complete
    await sleep(500);
};

export const ExampleCardboardListCalloutWithPrimaryButton = Template.bind(
    {}
) as TemplateStory;
ExampleCardboardListCalloutWithPrimaryButton.args = {
    ...getCommonDefaultProps(),
    listType: 'Complex',
    listItems: getDefaultCardboardListItems(),
    filterPredicate: (fakeItem: IFakeCardboardListItem, searchTerm) =>
        fakeItem.itemId.toLowerCase().includes(searchTerm.toLowerCase()),
    primaryActionProps: {
        primaryActionLabel: 'Create',
        onPrimaryActionClick: () => {
            console.log('Primary action button clicked!');
        }
    }
};
ExampleCardboardListCalloutWithPrimaryButton.play = async ({
    canvasElement
}) => {
    const canvas = within(canvasElement);

    // click on target to show callout
    const showCalloutButton = await canvas.findByTestId('showCalloutButton');
    await userEvent.click(showCalloutButton);

    // wait until callout show animation to complete
    await sleep(500);
};

export const ExampleCardboardBasicListCallout = Template.bind(
    {}
) as TemplateStory;
ExampleCardboardBasicListCallout.args = {
    ...getCommonDefaultProps(),
    listType: 'Basic',
    listItems: defaultCardboardBasicListItems,
    filterPredicate: (fakeItem: string, searchTerm) =>
        fakeItem.toLowerCase().includes(searchTerm.toLowerCase())
};
ExampleCardboardBasicListCallout.play = async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    // click on target to show callout
    const showCalloutButton = await canvas.findByTestId('showCalloutButton');
    await userEvent.click(showCalloutButton);

    // wait until callout show animation to complete
    await sleep(500);
};

export const ExampleCardboardBasicListCalloutWithPrimaryButton = Template.bind(
    {}
) as TemplateStory;
ExampleCardboardBasicListCalloutWithPrimaryButton.args = {
    ...getCommonDefaultProps(),
    listType: 'Basic',
    listItems: defaultCardboardBasicListItems,
    filterPredicate: (fakeItem: string, searchTerm) =>
        fakeItem.toLowerCase().includes(searchTerm.toLowerCase()),
    primaryActionProps: {
        primaryActionLabel: 'Create',
        onPrimaryActionClick: () => {
            console.log('Primary action button clicked!');
        }
    }
};
ExampleCardboardBasicListCalloutWithPrimaryButton.play = async ({
    canvasElement
}) => {
    const canvas = within(canvasElement);

    // click on target to show callout
    const showCalloutButton = await canvas.findByTestId('showCalloutButton');
    await userEvent.click(showCalloutButton);

    // wait until callout show animation to complete
    await sleep(500);
};
