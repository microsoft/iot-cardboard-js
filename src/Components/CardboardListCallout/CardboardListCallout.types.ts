import {
    DirectionalHint,
    ICalloutProps,
    IFocusZoneProps,
    IListProps
} from '@fluentui/react';
import { ICardboardListItem } from '../CardboardList/CardboardList.types';

export interface ICardboardListCalloutProps<T = void> {
    /** name of the class to put on the root node of the callout */
    className?: string;
    /** header of the callout */
    title: string;
    /** to pass if the list to be rendered is CardboardList or CardboardBasicList */
    listType: 'Complex' | 'Basic';
    /** unique identifier for this list of items. Will be joined with index */
    listKey: string;
    /** optional prop to set any specific list props needed for special cases */
    listProps?: Omit<IListProps, 'key' | 'items' | 'onRenderCell'>;
    /** optional prop to set any specific focus zone props needed for special cases */
    focusZoneProps?: Omit<IFocusZoneProps, 'isClickableOutsideFocusTrap'>;
    /** props needed for the callout */
    calloutProps?: Omit<
        ICalloutProps,
        'target' | 'isBeakVisible' | 'directionalHint' | 'onDismiss'
    >;
    /** the target id to position the callout */
    calloutTarget: string;
    /** how the callout should be positioned based on target element */
    directionalHint?: DirectionalHint;
    /** optional callback method when closing the callout */
    onDismiss?: () => void;
    /** collection of CardboardListItems or CardboardBasicListItems to include in the CardboardList or CardboardBasicList */
    listItems: (ICardboardListItem<T> | string)[];
    /** if list is being loaded (e.g. when data is pulled from adapter) */
    isListLoading?: boolean;
    /** text to be used as placeholder for the filter box */
    filterPlaceholder?: string;
    /** filter functiton to be used when a search term entered to filter list items */
    filterPredicate: (item: string | T, filterTerm: string) => void;
    /** data-testid to be set for the SearchBox component */
    searchBoxDataTestId?: string;
    /** text to be displayed when there is no item to display */
    noResultText: string;
    /** props for primary action button in the callout */
    primaryActionProps?: PrimaryActionProps;
}

type PrimaryActionProps = {
    primaryActionLabel: string;
    onPrimaryActionClick: () => void;
};
