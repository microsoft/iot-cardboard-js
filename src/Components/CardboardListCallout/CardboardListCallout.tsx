import React, { memo, useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
    DirectionalHint,
    FocusTrapCallout,
    PrimaryButton,
    SearchBox,
    Spinner,
    SpinnerSize,
    Stack,
    Text,
    useTheme
} from '@fluentui/react';
import { ICardboardListCalloutProps } from './CardboardListCallout.types';
import { ICardboardListItem } from '../CardboardList/CardboardList.types';
import { CardboardList } from '../CardboardList/CardboardList';
import {
    cardboardListCalloutPrimaryButtonStyles,
    getCardboardListCalloutComponentStyles,
    getCardboardListCalloutStyles
} from './CardboardListCallout.styles';
import { CardboardBasicList } from '../CardboardBasicList/CardboardBasicList';

/** This callout component consists a searchbox and list of items with an optional primary action */
const CardboardListCallout = <T extends unknown>({
    calloutProps,
    calloutTarget,
    className,
    dataButtonTestId,
    description,
    directionalHint,
    filterPlaceholder,
    filterPredicate,
    focusZoneProps,
    isListLoading,
    listItems,
    listKey,
    listProps,
    listType,
    noResultText,
    onDismiss,
    primaryActionProps,
    searchBoxDataTestId,
    title
}: ICardboardListCalloutProps<T>) => {
    const { t } = useTranslation();
    const [searchText, setSearchText] = useState('');
    const [filteredListItems, setFilteredListItems] = useState<
        Array<ICardboardListItem<T> | string>
    >(listItems);

    const filterListItems = useCallback(
        (searchTerm: string) => {
            setFilteredListItems(
                listItems.filter((listItem) =>
                    isCardboardListItem(listItem)
                        ? filterPredicate(listItem.item, searchTerm)
                        : filterPredicate(listItem, searchTerm)
                )
            );
        },
        [filterPredicate, listItems]
    );

    useEffect(() => {
        setFilteredListItems(listItems);
    }, [listItems]);

    const theme = useTheme();
    const styles = getCardboardListCalloutComponentStyles(theme);
    const calloutStyles = getCardboardListCalloutStyles(theme);
    return (
        <FocusTrapCallout
            className={className}
            {...calloutProps}
            focusTrapProps={{
                isClickableOutsideFocusTrap: true,
                ...focusZoneProps
            }}
            target={`#${calloutTarget}`}
            isBeakVisible={false}
            directionalHint={directionalHint ?? DirectionalHint.bottomLeftEdge}
            onDismiss={onDismiss}
            styles={calloutStyles}
        >
            <Stack tokens={{ childrenGap: 8 }}>
                <h4 className={styles.title}>{title}</h4>
                {description && (
                    <Text className={styles.description}>{description}</Text>
                )}
                {listItems.length > 0 && (
                    <SearchBox
                        {...(searchBoxDataTestId && {
                            'data-testid': searchBoxDataTestId
                        })}
                        placeholder={filterPlaceholder ?? t('search')}
                        onChange={(_event, value) => {
                            setSearchText(value);
                            filterListItems(value);
                        }}
                    />
                )}

                {isListLoading ? (
                    <Spinner size={SpinnerSize.xSmall} />
                ) : filteredListItems?.length === 0 ? (
                    <div className={styles.resultText}>{noResultText}</div>
                ) : listType === 'Basic' ? (
                    <CardboardBasicList
                        className={styles.list}
                        listProps={listProps}
                        items={filteredListItems as string[]}
                        listKey={listKey}
                        textToHighlight={searchText}
                    />
                ) : (
                    <CardboardList<T>
                        className={styles.list}
                        listProps={listProps}
                        items={filteredListItems as ICardboardListItem<T>[]}
                        listKey={listKey}
                        textToHighlight={searchText}
                    />
                )}

                {primaryActionProps && (
                    <PrimaryButton
                        data-testid={dataButtonTestId}
                        styles={cardboardListCalloutPrimaryButtonStyles}
                        onClick={primaryActionProps.onPrimaryActionClick}
                        disabled={primaryActionProps.disabled}
                    >
                        {primaryActionProps.primaryActionLabel}
                    </PrimaryButton>
                )}
            </Stack>
        </FocusTrapCallout>
    );
};

const isCardboardListItem = (
    listItem: any
): listItem is ICardboardListItem<any> => listItem.item;

export default memo(CardboardListCallout);
