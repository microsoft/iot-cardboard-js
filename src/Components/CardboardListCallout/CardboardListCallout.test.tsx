import React from 'react';
import { DirectionalHint } from '@fluentui/react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import user from '@testing-library/user-event';
import { ICardboardListItem } from '../CardboardList/CardboardList.types';
import CardboardListCallout from './CardboardListCallout';
import { useId } from '@fluentui/react-hooks';

interface IFakeCardboardListItem {
    itemId: string;
    itemDescription: string | undefined;
}
const defaultCardboardListItems: IFakeCardboardListItem[] = [
    {
        itemId: 'Item 1',
        itemDescription: 'description for item 1'
    },
    {
        itemId: 'Item 2',
        itemDescription: 'description for item 1'
    },
    {
        itemId: 'Item 3',
        itemDescription: 'description for item 3'
    },
    {
        itemId: 'Item 4',
        itemDescription: 'description for item 4'
    },
    {
        itemId: 'Item 5',
        itemDescription: 'description for item 5'
    }
];

const defaultOnClickHandler = (item) => {
    alert(`clicked item ${item.itemId}`);
};

const getDefaultCardboardListItems = (): Array<
    ICardboardListItem<IFakeCardboardListItem>
> =>
    defaultCardboardListItems.map((item) => ({
        ariaLabel: '',
        iconStart: { name: 'Shapes' },
        iconEnd: { name: 'Add' },
        item: item,
        onClick: defaultOnClickHandler,
        textPrimary: item.itemId,
        textSecondary: item.itemDescription
    }));

interface TestComponentProps {
    onPrimaryActionClick: (searchTerm) => void;
}
const TestComponent: React.FC<TestComponentProps> = ({
    onPrimaryActionClick
}) => {
    const testCalloutTargetId = useId('test-cardboard-list-callout-target');

    return (
        <CardboardListCallout
            title="Test cardboard list callout"
            listKey="testList"
            directionalHint={DirectionalHint.bottomLeftEdge}
            filterPlaceholder="Search"
            noResultText="No result"
            listType="Complex"
            listItems={getDefaultCardboardListItems() as any}
            filterPredicate={(fakeItem, searchTerm) =>
                (fakeItem as IFakeCardboardListItem).itemId
                    .toLowerCase()
                    .includes(searchTerm.toLowerCase())
            }
            calloutTarget={testCalloutTargetId}
            onDismiss={() => {
                console.log('Dismissed');
            }}
            primaryActionProps={{
                primaryActionLabel: 'Create',
                onPrimaryActionClick: onPrimaryActionClick
            }}
            dataButtonTestId="create-item-callout-button"
            searchBoxDataTestId="search-box"
            focusTrapTestId="focus-trap"
        />
    );
};

test('entered search term passed successfully on primary button click', async () => {
    // setup a DOM element as a render target
    const testCallback = jest.fn();
    const { getByTestId } = render(
        <TestComponent onPrimaryActionClick={testCallback} />
    );

    const modalComponent = screen.getByTestId('focus-trap'); // take the snapshot of the callout modal component with search text
    expect(modalComponent).toMatchSnapshot();

    const searchTerm = 'test';
    const searchBox = getByTestId('search-box');
    fireEvent.change(searchBox, { target: { value: searchTerm } }); // emulate the user typing input

    const createNewButton = getByTestId('create-item-callout-button'); // click on the primary button in the callout
    await waitFor(() => user.click(createNewButton));

    expect(testCallback).toHaveBeenCalledWith(searchTerm); // test if the search term is successfully passed to the on primary button click event handler method through component's callback
});
