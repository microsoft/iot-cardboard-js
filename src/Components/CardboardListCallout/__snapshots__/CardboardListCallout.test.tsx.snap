// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardboardListCallout entered search term passed successfully on primary button click 1`] = `
<div
  class="ms-Callout root-131"
  data-testid="focus-trap"
  style="opacity: 0; filter: opacity(0); pointer-events: none;"
  tabindex="-1"
>
  <div
    class="ms-Callout-main calloutMain-134"
    style="outline: none; max-height: 100%;"
  >
    <div>
      <div
        aria-hidden="true"
        data-is-focus-trap-zone-bumper="true"
        data-is-visible="true"
        style="pointer-events: none; position: fixed;"
        tabindex="0"
      />
      <div
        class="ms-Stack css-135"
      >
        <h4
          class="cb-list-callout-title title-121"
        >
          Test cardboard list callout
        </h4>
        <div
          class="ms-SearchBox root-136"
        >
          <div
            aria-hidden="true"
            class="ms-SearchBox-iconContainer iconContainer-137"
          >
            <i
              aria-hidden="true"
              class="ms-SearchBox-icon icon-143"
              data-icon-name="Search"
            >
              
            </i>
          </div>
          <input
            class="ms-SearchBox-field field-140"
            data-testid="search-box"
            id="SearchBox2"
            placeholder="Search"
            role="searchbox"
            value=""
          />
        </div>
        <div
          class="ms-FocusZone css-144 cb-list-callout-list list-123"
          data-focuszone-id="FocusZone3"
        >
          <div
            class="ms-List"
            data-testid="cardboard-list-testList"
            role="list"
          >
            <div
              class="ms-List-surface"
              role="presentation"
            >
              <div
                class="ms-List-page"
                role="presentation"
              >
                <div
                  class="ms-List-cell"
                  data-automationid="ListCell"
                  data-list-index="0"
                  role="listitem"
                >
                  <span
                    class="cb-list-item-root root-162"
                  >
                    <button
                      class="ms-Button ms-Button--default root-165"
                      data-is-focusable="true"
                      data-testid="cardboard-list-item-testList-0"
                      type="button"
                    >
                      <span
                        class="ms-Button-flexContainer flexContainer-166"
                        data-automationid="splitbuttonprimary"
                      >
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-168 cb-list-item-icon icon-156"
                          data-icon-name="Shapes"
                          style="font-family: \\"FabricMDL2Icons-5\\";"
                        >
                          
                        </i>
                        <div
                          class="cb-list-item-text-container textContainer-160"
                        >
                          <div
                            class="cb-list-item-primary-text primaryText-161"
                            title="Item 1"
                          >
                            Item 1
                          </div>
                          <div
                            class="cb-list-item-secondary-text secondaryText-163"
                            title="description for item 1"
                          >
                            description for item 1
                          </div>
                        </div>
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-142 cb-list-item-icon icon-156 cb-list-item-end-icon endIcon-155"
                          data-icon-name="Add"
                          style="font-family: \\"FabricMDL2Icons\\";"
                        >
                          
                        </i>
                      </span>
                    </button>
                  </span>
                </div>
                <div
                  class="ms-List-cell"
                  data-automationid="ListCell"
                  data-list-index="1"
                  role="listitem"
                >
                  <span
                    class="cb-list-item-root root-162"
                  >
                    <button
                      class="ms-Button ms-Button--default root-165"
                      data-is-focusable="true"
                      data-testid="cardboard-list-item-testList-1"
                      type="button"
                    >
                      <span
                        class="ms-Button-flexContainer flexContainer-166"
                        data-automationid="splitbuttonprimary"
                      >
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-168 cb-list-item-icon icon-156"
                          data-icon-name="Shapes"
                          style="font-family: \\"FabricMDL2Icons-5\\";"
                        >
                          
                        </i>
                        <div
                          class="cb-list-item-text-container textContainer-160"
                        >
                          <div
                            class="cb-list-item-primary-text primaryText-161"
                            title="Item 2"
                          >
                            Item 2
                          </div>
                          <div
                            class="cb-list-item-secondary-text secondaryText-163"
                            title="description for item 1"
                          >
                            description for item 1
                          </div>
                        </div>
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-142 cb-list-item-icon icon-156 cb-list-item-end-icon endIcon-155"
                          data-icon-name="Add"
                          style="font-family: \\"FabricMDL2Icons\\";"
                        >
                          
                        </i>
                      </span>
                    </button>
                  </span>
                </div>
                <div
                  class="ms-List-cell"
                  data-automationid="ListCell"
                  data-list-index="2"
                  role="listitem"
                >
                  <span
                    class="cb-list-item-root root-162"
                  >
                    <button
                      class="ms-Button ms-Button--default root-165"
                      data-is-focusable="true"
                      data-testid="cardboard-list-item-testList-2"
                      type="button"
                    >
                      <span
                        class="ms-Button-flexContainer flexContainer-166"
                        data-automationid="splitbuttonprimary"
                      >
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-168 cb-list-item-icon icon-156"
                          data-icon-name="Shapes"
                          style="font-family: \\"FabricMDL2Icons-5\\";"
                        >
                          
                        </i>
                        <div
                          class="cb-list-item-text-container textContainer-160"
                        >
                          <div
                            class="cb-list-item-primary-text primaryText-161"
                            title="Item 3"
                          >
                            Item 3
                          </div>
                          <div
                            class="cb-list-item-secondary-text secondaryText-163"
                            title="description for item 3"
                          >
                            description for item 3
                          </div>
                        </div>
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-142 cb-list-item-icon icon-156 cb-list-item-end-icon endIcon-155"
                          data-icon-name="Add"
                          style="font-family: \\"FabricMDL2Icons\\";"
                        >
                          
                        </i>
                      </span>
                    </button>
                  </span>
                </div>
                <div
                  class="ms-List-cell"
                  data-automationid="ListCell"
                  data-list-index="3"
                  role="listitem"
                >
                  <span
                    class="cb-list-item-root root-162"
                  >
                    <button
                      class="ms-Button ms-Button--default root-165"
                      data-is-focusable="true"
                      data-testid="cardboard-list-item-testList-3"
                      type="button"
                    >
                      <span
                        class="ms-Button-flexContainer flexContainer-166"
                        data-automationid="splitbuttonprimary"
                      >
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-168 cb-list-item-icon icon-156"
                          data-icon-name="Shapes"
                          style="font-family: \\"FabricMDL2Icons-5\\";"
                        >
                          
                        </i>
                        <div
                          class="cb-list-item-text-container textContainer-160"
                        >
                          <div
                            class="cb-list-item-primary-text primaryText-161"
                            title="Item 4"
                          >
                            Item 4
                          </div>
                          <div
                            class="cb-list-item-secondary-text secondaryText-163"
                            title="description for item 4"
                          >
                            description for item 4
                          </div>
                        </div>
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-142 cb-list-item-icon icon-156 cb-list-item-end-icon endIcon-155"
                          data-icon-name="Add"
                          style="font-family: \\"FabricMDL2Icons\\";"
                        >
                          
                        </i>
                      </span>
                    </button>
                  </span>
                </div>
                <div
                  class="ms-List-cell"
                  data-automationid="ListCell"
                  data-list-index="4"
                  role="listitem"
                >
                  <span
                    class="cb-list-item-root root-162"
                  >
                    <button
                      class="ms-Button ms-Button--default root-165"
                      data-is-focusable="true"
                      data-testid="cardboard-list-item-testList-4"
                      type="button"
                    >
                      <span
                        class="ms-Button-flexContainer flexContainer-166"
                        data-automationid="splitbuttonprimary"
                      >
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-168 cb-list-item-icon icon-156"
                          data-icon-name="Shapes"
                          style="font-family: \\"FabricMDL2Icons-5\\";"
                        >
                          
                        </i>
                        <div
                          class="cb-list-item-text-container textContainer-160"
                        >
                          <div
                            class="cb-list-item-primary-text primaryText-161"
                            title="Item 5"
                          >
                            Item 5
                          </div>
                          <div
                            class="cb-list-item-secondary-text secondaryText-163"
                            title="description for item 5"
                          >
                            description for item 5
                          </div>
                        </div>
                        <i
                          aria-hidden="true"
                          class="ms-Icon root-105 css-142 cb-list-item-icon icon-156 cb-list-item-end-icon endIcon-155"
                          data-icon-name="Add"
                          style="font-family: \\"FabricMDL2Icons\\";"
                        >
                          
                        </i>
                      </span>
                    </button>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          class="ms-Button ms-Button--primary cb-list-callout-primary-button primaryButton-145"
          data-is-focusable="true"
          data-testid="create-item-callout-button"
          type="button"
        >
          <span
            class="ms-Button-flexContainer flexContainer-146"
            data-automationid="splitbuttonprimary"
          >
            <span
              class="ms-Button-textContainer textContainer-147"
            >
              <span
                class="ms-Button-label label-149"
                id="id__4"
              >
                Create
              </span>
            </span>
          </span>
        </button>
      </div>
      <div
        aria-hidden="true"
        data-is-focus-trap-zone-bumper="true"
        data-is-visible="true"
        style="pointer-events: none; position: fixed;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
